File: scripts/database/reset-database-complete.mjs
/**
 * Complete Database Reset Script
 *
 * This script:
 * 1. Deletes knowledge base data (chunks, embeddings, products, images, etc.)
 * 2. PRESERVES user data (users, profiles, workspaces, API keys)
 * 3. Deletes all files from storage buckets EXCEPT pdf-documents folder
 * 4. Verifies cleanup was successful
 * 5. Reports storage and resource usage
 *
 * ⚠️ PRESERVED DATA:
 * - Users and authentication (auth.users)
 * - User profiles
 * - Workspaces and workspace members
 * - API keys and usage logs
 * - PDF documents in pdf-documents folder (ALL FILES PRESERVED)
 *
 * 🗑️ DELETED DATA:
 * - All PDF processing data (chunks, embeddings, images)
 * - All products and materials
 * - All background jobs and processing results
 * - All analytics and agent tasks
 * - All 3D generation history
 * - Storage files (except pdf-documents folder)
 */

import fetch from 'node-fetch';

const SUPABASE_URL = 'https://bgbavxtjlbvgplozizxu.supabase.co';
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_SERVICE_ROLE_KEY) {
  console.error('❌ SUPABASE_SERVICE_ROLE_KEY environment variable is required');
  process.exit(1);
}

// Tables to clear (in order to respect foreign key constraints)
const TABLES_TO_CLEAR = [
  'embeddings',
  'document_images',
  'document_chunks',
  'products',
  'background_jobs',
  'documents',
  'ai_analysis_queue',
  'processed_documents',
  'job_progress',
  'materials_catalog',
  'material_visual_analysis',
  'processing_results',
  'quality_metrics_daily',
  'quality_scoring_logs',
  'analytics_events',
  'agent_tasks',
  'generation_3d',
  'scraped_materials_temp',
  'scraping_sessions',
  'scraping_pages'
];

// Storage buckets configuration
const BUCKETS_CONFIG = [
  { name: 'pdf-tiles', excludeFolders: [] },
  { name: 'pdf-documents', excludeFolders: ['*'] },  // ⚠️ PRESERVE ALL files
  { name: 'material-images', excludeFolders: [] }
];

async function makeSupabaseRequest(method, path, body = null) {
  const url = `${SUPABASE_URL}${path}`;
  const options = {
    method,
    headers: {
      'Authorization': `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
      'apikey': SUPABASE_SERVICE_ROLE_KEY,
      'Prefer': 'return=minimal'
    }
  };

  if (body) {
    options.headers['Content-Type'] = 'application/json';
    options.body = JSON.stringify(body);
  }

  const response = await fetch(url, options);

  if (method === 'DELETE' && (response.status === 200 || response.status === 204)) {
    return { success: true };
  }

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Supabase API error ${response.status}: ${text}`);
  }

  const text = await response.text();
  if (!text) return { success: true };

  try {
    return JSON.parse(text);
  } catch (e) {
    return { success: true, raw: text };
  }
}

async function clearTable(tableName) {
  console.log(`\n🗑️  Clearing table: ${tableName}`);

  try {
    const countBefore = await makeSupabaseRequest('GET', `/rest/v1/${tableName}?select=count`, null);
    const count = countBefore?.[0]?.count || 0;

    if (count === 0) {
      console.log(`   ✅ Table ${tableName} is already empty`);
      return { table: tableName, deleted: 0 };
    }

    console.log(`   📊 Found ${count} rows to delete`);
    await makeSupabaseRequest('DELETE', `/rest/v1/${tableName}?id=neq.00000000-0000-0000-0000-000000000000`, null);
    console.log(`   ✅ Deleted ${count} rows from ${tableName}`);
    return { table: tableName, deleted: count };
  } catch (error) {
    console.error(`   ❌ Failed to clear ${tableName}: ${error.message}`);
    return { table: tableName, deleted: 0, error: error.message };
  }
}

async function listBucketFiles(bucketName, path = '') {
  try {
    const response = await makeSupabaseRequest('POST', `/storage/v1/object/list/${bucketName}`, {
      prefix: path,
      limit: 1000,
      offset: 0
    });
    return response || [];
  } catch (error) {
    console.error(`   ❌ Failed to list files in ${bucketName}/${path}: ${error.message}`);
    return [];
  }
}

async function deleteFile(bucketName, filePath) {
  try {
    await makeSupabaseRequest('DELETE', `/storage/v1/object/${bucketName}`, {
      prefixes: [filePath]
    });
    return true;
  } catch (error) {
    console.error(`   ⚠️  Failed to delete ${bucketName}/${filePath}: ${error.message}`);
    return false;
  }
}

async function listAllFilesRecursively(bucketName, prefix = '') {
  const allFiles = [];

  async function listFolder(folderPath) {
    const items = await listBucketFiles(bucketName, folderPath);

    for (const item of items) {
      const fullPath = folderPath ? `${folderPath}/${item.name}` : item.name;

      if (!item.metadata || item.metadata.mimetype === 'application/x-directory') {
        await listFolder(fullPath);
      } else {
        allFiles.push(fullPath);
      }
    }
  }

  await listFolder(prefix);
  return allFiles;
}

async function clearBucket(bucketConfig) {
  const { name: bucketName, excludeFolders = [] } = bucketConfig;
  console.log(`\n🗑️  Clearing bucket: ${bucketName}`);

  if (excludeFolders.length > 0) {
    if (excludeFolders[0] === '*') {
      console.log(`   🔒 PRESERVING ALL FILES in ${bucketName}`);
    } else {
      console.log(`   🔒 Preserving folders: ${excludeFolders.join(', ')}`);
    }
  }

  try {
    const allFiles = await listAllFilesRecursively(bucketName);

    if (allFiles.length === 0) {
      console.log(`   ✅ Bucket ${bucketName} is already empty`);
      return { bucket: bucketName, deleted: 0, skipped: 0 };
    }

    console.log(`   📊 Found ${allFiles.length} files to process`);

    let deleted = 0;
    let failed = 0;
    let skipped = 0;

    for (const filePath of allFiles) {
      let shouldSkip = false;
      for (const excludedFolder of excludeFolders) {
        if (excludedFolder === '*' || filePath.startsWith(excludedFolder)) {
          shouldSkip = true;
          skipped++;
          break;
        }
      }

      if (shouldSkip) continue;

      const success = await deleteFile(bucketName, filePath);
      if (success) {
        deleted++;
        if (deleted % 10 === 0) {
          console.log(`   🔄 Deleted ${deleted} files...`);
        }
      } else {
        failed++;
      }
    }

    console.log(`   ✅ Deleted ${deleted} files from ${bucketName}`);
    if (skipped > 0) {
      console.log(`   🔒 Preserved ${skipped} files`);
    }
    if (failed > 0) {
      console.log(`   ⚠️  Failed to delete ${failed} files`);
    }

    return { bucket: bucketName, deleted, failed, skipped };
  } catch (error) {
    console.error(`   ❌ Failed to clear bucket ${bucketName}: ${error.message}`);
    return { bucket: bucketName, deleted: 0, failed: 0, skipped: 0, error: error.message };
  }
}

async function main() {
  console.log('═══════════════════════════════════════════════════════════');
  console.log('🔄 COMPLETE DATABASE & FILES CLEANUP');
  console.log('═══════════════════════════════════════════════════════════');
  console.log('');
  console.log('✅ PRESERVED:');
  console.log('   • Users & Authentication');
  console.log('   • Profiles & Workspaces');
  console.log('   • API Keys & Usage Logs');
  console.log('   • ALL PDF files in pdf-documents folder');
  console.log('');
  console.log('🗑️  WILL DELETE:');
  console.log('   • PDF Processing Data (chunks, embeddings, images)');
  console.log('   • Products & Materials Catalog');
  console.log('   • Background Jobs & Processing Results');
  console.log('   • Analytics & Agent Tasks');
  console.log('   • Storage files (except pdf-documents folder)');
  console.log('');
  console.log('═══════════════════════════════════════════════════════════');
  console.log(`📅 Started: ${new Date().toISOString()}\n`);

  const results = {
    tables: [],
    buckets: []
  };

  // Step 1: Clear database tables
  console.log('\n🗑️  STEP 1: Clear knowledge base tables');
  console.log(`   📋 Clearing ${TABLES_TO_CLEAR.length} tables...`);
  for (const tableName of TABLES_TO_CLEAR) {
    const result = await clearTable(tableName);
    results.tables.push(result);
  }

  // Step 2: Clear storage buckets
  console.log('\n🗑️  STEP 2: Clear storage buckets');
  for (const bucketConfig of BUCKETS_CONFIG) {
    const result = await clearBucket(bucketConfig);
    results.buckets.push(result);
  }

  // Summary
  console.log('\n═══════════════════════════════════════════════════════════');
  console.log('📊 CLEANUP SUMMARY');
  console.log('═══════════════════════════════════════════════════════════');

  const totalRowsDeleted = results.tables.reduce((sum, r) => sum + (r.deleted || 0), 0);
  const totalFilesDeleted = results.buckets.reduce((sum, r) => sum + (r.deleted || 0), 0);
  const totalFilesSkipped = results.buckets.reduce((sum, r) => sum + (r.skipped || 0), 0);

  console.log(`\n✅ Database rows deleted: ${totalRowsDeleted}`);
  console.log(`✅ Storage files deleted: ${totalFilesDeleted}`);
  console.log(`🔒 Storage files preserved: ${totalFilesSkipped} (pdf-documents folder)`);

  console.log('\n📋 Table cleanup details:');
  console.table(results.tables);

  console.log('\n📦 Bucket cleanup details:');
  console.table(results.buckets);

  console.log(`\n📅 Completed: ${new Date().toISOString()}`);
  console.log('═══════════════════════════════════════════════════════════');
}

main().catch(error => {
  console.error('\n❌ FATAL ERROR:', error);
  process.exit(1);
});

File: scripts/database/reset_database_complete.py
#!/usr/bin/env python3
"""
Complete Database Reset Script (Python Version)

This script performs a comprehensive cleanup of all user-generated data while preserving
system configuration and authentication data.

DELETES:
1. Knowledge base data (chunks, embeddings, products, images, etc.)
2. Agent chat conversations and messages
3. CRM contacts and relationships
4. All jobs (background_jobs, job_checkpoints, job_progress)
5. Moodboards and moodboard items
6. 3D generation history
7. All analytics data
8. All metadata values and relevancy relationships
9. All quotes and quote items
10. All files from storage buckets EXCEPT pdf-documents folder

PRESERVES:
1. User data (users, profiles, workspaces, API keys)
2. Global upsells (admin-managed upsell items)
3. Global timeline elements (timeline_steps)
4. Material metadata field definitions (schema)
5. PDF files in pdf-documents folder
6. System settings and configuration
"""

import os
import sys
import requests
from datetime import datetime

SUPABASE_URL = 'https://bgbavxtjlbvgplozizxu.supabase.co'
SUPABASE_SERVICE_ROLE_KEY = os.environ.get('SUPABASE_SERVICE_ROLE_KEY')

if not SUPABASE_SERVICE_ROLE_KEY:
    print('❌ SUPABASE_SERVICE_ROLE_KEY environment variable is required')
    sys.exit(1)

# Tables to clear (in order to respect foreign key constraints)
# Order matters! Delete child tables before parent tables
TABLES_TO_CLEAR = [
    # Agent Chat System (DELETE)
    'agent_chat_messages',           # Chat messages (child of conversations)
    'agent_chat_conversations',      # Chat conversations
    'agent_uploaded_files',          # Files uploaded in chat

    # CRM Contacts (DELETE)
    'crm_contact_relationships',     # Contact relationships (child of contacts)
    'crm_contacts',                  # CRM contacts

    # Quotes System (DELETE - except global upsells and timeline steps)
    'quote_timeline',                # Quote timeline progress (child of quotes)
    'quote_upsells',                 # Quote upsells junction (child of quotes)
    'quote_items',                   # Quote items (child of quotes)
    'quotes',                        # Quotes
    'status_tags',                   # Custom status tags
    # NOTE: 'upsells' and 'timeline_steps' are PRESERVED (global data)

    # Moodboards (DELETE)
    'moodboard_quote_requests',      # Moodboard quote requests (child of moodboards)
    'moodboard_products',            # Moodboard products (child of moodboards)
    'moodboard_items',               # Moodboard items (child of moodboards)
    'moodboards',                    # Moodboards

    # 3D Generation (DELETE)
    'generation_3d',                 # 3D generation history

    # Analytics (DELETE)
    'analytics_events',              # Analytics events
    'quality_metrics_daily',         # Daily quality metrics
    'quality_scoring_logs',          # Quality scoring logs
    'recommendation_analytics',      # Recommendation analytics (if exists)

    # Document Entities & Relationships (DELETE)
    'product_document_relationships', # Product-document entity relationships
    'document_entities',             # Document entities (certificates, logos, specs)

    # Relevancy Relationships (DELETE)
    'product_chunk_relationships',   # Product-chunk relevancies
    'chunk_image_relationships',     # Chunk-image relevancies
    'product_image_relationships',   # Product-image relevancies

    # Metadata (DELETE)
    'metafield_values',              # Metafield values (child of metadata fields)
    # NOTE: 'material_metadata_fields' is PRESERVED (schema definition)

    # PDF Processing & Knowledge Base (DELETE)
    'job_checkpoints',               # Job checkpoints (child of background_jobs)
    'job_progress',                  # Job progress tracking (child of background_jobs)
    'ai_analysis_queue',             # AI analysis queue
    'image_processing_queue',        # Image processing queue
    'embeddings',                    # Text and image embeddings
    'document_images',               # Extracted images from PDFs
    'document_chunks',               # Semantic text chunks
    'products',                      # Extracted products
    'background_jobs',               # Processing jobs
    'documents',                     # PDF documents metadata
    'processed_documents',           # Processed document records

    # Materials & Catalog (DELETE)
    'materials_catalog',             # Materials catalog entries
    'material_visual_analysis',      # Visual analysis results

    # Processing & Quality (DELETE)
    'processing_results',            # Processing results

    # Agent Tasks (DELETE)
    'agent_tasks',                   # Agent task records

    # Web Scraping (DELETE)
    'scraped_materials_temp',        # Temporary scraped materials
    'scraping_sessions',             # Scraping sessions
    'scraping_pages',                # Scraping pages

    # Data Import (DELETE)
    'data_import_jobs',              # Data import jobs (if exists)
    'data_import_history',           # Data import history (if exists)
]

# Storage buckets configuration
BUCKETS_CONFIG = [
    {'name': 'pdf-tiles', 'exclude_folders': []},
    {'name': 'pdf-documents', 'exclude_folders': ['*']},  # Preserve all files
    {'name': 'material-images', 'exclude_folders': []}
]


def make_supabase_request(method, path, body=None):
    """Make a request to Supabase API"""
    url = f"{SUPABASE_URL}{path}"
    headers = {
        'Authorization': f'Bearer {SUPABASE_SERVICE_ROLE_KEY}',
        'apikey': SUPABASE_SERVICE_ROLE_KEY,
        'Prefer': 'return=minimal'
    }
    
    if body:
        headers['Content-Type'] = 'application/json'
        response = requests.request(method, url, headers=headers, json=body)
    else:
        response = requests.request(method, url, headers=headers)
    
    # For DELETE requests, 200 or 204 is OK
    if method == 'DELETE' and response.status_code in [200, 204]:
        return {'success': True}
    
    if not response.ok:
        raise Exception(f"Supabase API error {response.status_code}: {response.text}")
    
    # Handle empty responses
    if not response.text:
        return {'success': True}
    
    try:
        return response.json()
    except:
        return {'success': True, 'raw': response.text}


def clear_table(table_name):
    """Clear all rows from a table"""
    print(f"\n🗑️  Clearing table: {table_name}")
    
    try:
        # Get count before deletion
        count_response = make_supabase_request('GET', f'/rest/v1/{table_name}?select=count')
        count = count_response[0].get('count', 0) if count_response else 0
        
        if count == 0:
            print(f"   ✅ Table {table_name} is already empty")
            return {'table': table_name, 'deleted': 0}
        
        print(f"   📊 Found {count} rows to delete")
        
        # Delete all rows
        make_supabase_request('DELETE', f'/rest/v1/{table_name}?id=neq.00000000-0000-0000-0000-000000000000')
        
        print(f"   ✅ Deleted {count} rows from {table_name}")
        return {'table': table_name, 'deleted': count}
    except Exception as error:
        print(f"   ❌ Failed to clear {table_name}: {error}")
        return {'table': table_name, 'deleted': 0, 'error': str(error)}


def list_bucket_files(bucket_name, path=''):
    """List files in a storage bucket"""
    try:
        response = make_supabase_request('POST', f'/storage/v1/object/list/{bucket_name}', {
            'prefix': path,
            'limit': 1000,
            'offset': 0
        })
        return response or []
    except Exception as error:
        print(f"   ❌ Failed to list files in {bucket_name}/{path}: {error}")
        return []


def delete_file(bucket_name, file_path):
    """Delete a file from storage"""
    try:
        make_supabase_request('DELETE', f'/storage/v1/object/{bucket_name}', {
            'prefixes': [file_path]
        })
        return True
    except Exception as error:
        print(f"   ⚠️  Failed to delete {bucket_name}/{file_path}: {error}")
        return False


def list_all_files_recursively(bucket_name, prefix=''):
    """List all files in a bucket recursively"""
    all_files = []

    def list_folder(folder_path):
        items = list_bucket_files(bucket_name, folder_path)

        for item in items:
            full_path = f"{folder_path}/{item['name']}" if folder_path else item['name']

            # If it's a folder, recurse into it
            if not item.get('metadata') or item.get('metadata', {}).get('mimetype') == 'application/x-directory':
                list_folder(full_path)
            else:
                # It's a file, add it to the list
                all_files.append(full_path)

    list_folder(prefix)
    return all_files


def clear_bucket(bucket_config):
    """Clear files from a storage bucket"""
    bucket_name = bucket_config['name']
    exclude_folders = bucket_config.get('exclude_folders', [])

    print(f"\n🗑️  Clearing bucket: {bucket_name}")

    if exclude_folders:
        print(f"   🔒 Preserving folders: {', '.join(exclude_folders)}")

    try:
        # List all files recursively
        all_files = list_all_files_recursively(bucket_name)

        if not all_files:
            print(f"   ✅ Bucket {bucket_name} is already empty")
            return {'bucket': bucket_name, 'deleted': 0, 'skipped': 0}

        print(f"   📊 Found {len(all_files)} files to process")

        deleted = 0
        failed = 0
        skipped = 0

        # Delete files in batches
        for file_path in all_files:
            # Check if file is in an excluded folder
            should_skip = False
            for excluded_folder in exclude_folders:
                # Special case: '*' means preserve ALL files in this bucket
                if excluded_folder == '*' or file_path.startswith(excluded_folder):
                    should_skip = True
                    skipped += 1
                    break

            if should_skip:
                if skipped % 10 == 0 and skipped > 0:
                    print(f"   🔒 Skipped {skipped} files in preserved folders...")
                continue

            success = delete_file(bucket_name, file_path)
            if success:
                deleted += 1
                if deleted % 10 == 0:
                    print(f"   🔄 Deleted {deleted} files...")
            else:
                failed += 1

        print(f"   ✅ Deleted {deleted} files from {bucket_name}")
        if skipped > 0:
            print(f"   🔒 Preserved {skipped} files in excluded folders")
        if failed > 0:
            print(f"   ⚠️  Failed to delete {failed} files")

        return {'bucket': bucket_name, 'deleted': deleted, 'failed': failed, 'skipped': skipped}
    except Exception as error:
        print(f"   ❌ Failed to clear bucket {bucket_name}: {error}")
        return {'bucket': bucket_name, 'deleted': 0, 'failed': 0, 'skipped': 0, 'error': str(error)}


def main():
    """Main cleanup function"""
    print('═' * 100)
    print('🔄 COMPLETE DATABASE RESET (PRESERVING USER DATA & SYSTEM CONFIG)')
    print('═' * 100)
    print('')
    print('✅ PRESERVED:')
    print('   • Users & Authentication')
    print('   • Profiles & Workspaces')
    print('   • API Keys & Usage Logs')
    print('   • Global Upsells (admin-managed)')
    print('   • Global Timeline Elements (timeline_steps)')
    print('   • Material Metadata Field Definitions (schema)')
    print('   • PDF files in pdf-documents folder')
    print('   • System Settings & Configuration')
    print('')
    print('🗑️  WILL DELETE:')
    print('   • Agent Chat Conversations & Messages')
    print('   • CRM Contacts & Relationships')
    print('   • All Jobs (background_jobs, checkpoints, progress)')
    print('   • Moodboards & Moodboard Items')
    print('   • 3D Generation History')
    print('   • All Analytics Data')
    print('   • All Metadata Values & Relevancy Relationships')
    print('   • All Quotes & Quote Items')
    print('   • PDF Processing Data (chunks, embeddings, images)')
    print('   • Products & Materials Catalog')
    print('   • Storage files (except pdf-documents folder)')
    print('')
    print('═' * 100)
    print(f'📅 Started: {datetime.now().isoformat()}')

    results = {
        'tables': [],
        'buckets': []
    }

    # Step 1: Clear database tables
    print('\n🗑️  STEP 1: Clear knowledge base tables')
    print(f'   📋 Clearing {len(TABLES_TO_CLEAR)} tables (preserving user data)...')
    for table_name in TABLES_TO_CLEAR:
        result = clear_table(table_name)
        results['tables'].append(result)

    # Step 2: Clear storage buckets
    print('\n🗑️  STEP 2: Clear storage buckets')
    print('   🔒 Preserving pdf-documents folder and all files inside...')
    for bucket_config in BUCKETS_CONFIG:
        result = clear_bucket(bucket_config)
        results['buckets'].append(result)

    # Summary
    print('\n' + '═' * 100)
    print('📊 CLEANUP SUMMARY')
    print('═' * 100)

    total_rows_deleted = sum(r.get('deleted', 0) for r in results['tables'])
    total_files_deleted = sum(r.get('deleted', 0) for r in results['buckets'])
    total_files_failed = sum(r.get('failed', 0) for r in results['buckets'])
    total_files_skipped = sum(r.get('skipped', 0) for r in results['buckets'])

    print(f"\n✅ Database rows deleted: {total_rows_deleted}")
    print(f"✅ Storage files deleted: {total_files_deleted}")
    if total_files_skipped > 0:
        print(f"🔒 Storage files preserved: {total_files_skipped} (pdf-documents folder)")
    if total_files_failed > 0:
        print(f"⚠️  Storage files failed: {total_files_failed}")

    print('\n✅ PRESERVED DATA:')
    print('   • Users, Profiles, Workspaces remain intact')
    print('   • API Keys and authentication preserved')
    print('   • Global Upsells (admin-managed upsell items)')
    print('   • Global Timeline Elements (timeline_steps)')
    print('   • Material Metadata Field Definitions (schema)')
    print(f'   • {total_files_skipped} files preserved in pdf-documents folder')
    print('   • System Settings & Configuration')

    print(f"\n📅 Completed: {datetime.now().isoformat()}")
    print('═' * 100)


if __name__ == '__main__':
    main()


